### PR TITLE
Add analyzer/fixer to suggest changing code fluent collection building to a collection expression

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Part of https://github.com/dotnet/roslyn/pull/69473

Finds code like: `ImmutableArray<int>.Empty.AddRange(x).AddRange(y).ToXXX()` (and many patterns like that) and offers to convert to code like `[..y, ..y]`.

This works for varying patterns of `Add/AddRange` as well as `Concat` (on a reified list).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@akhera99 for review.  thanks!

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+                if (matchesInReverse != null)
+                {
+                    AddArguments(matchesInReverse, GetArguments(currentInvocationExpression, unwrapArgument), useSpread: false);
+                }
+
+                return true;
+            }
+
+            // Down to some final collection.  Like `x` in `x.Concat(y).ToArray()`.  If `x` is itself is something that
+            // can be iterated, we can conver this to `[.. x, .. y]`.
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer.cs:221]:
```suggestion
            // can be iterated, we can convert this to `[.. x, .. y]`.
```

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs [hunk 2]
```diff
@@ -91,7 +91,12 @@ internal static class UseCollectionExpressionHelpers
         // Ensure that we have a collection conversion with the replacement.  If not, this wasn't a legal replacement
         // (for example, we're trying to replace an expression that is converted to something that isn't even a
         // collection type).
+        //
+        // Note: an identity conversion is always legal without needing any more checks.
         var conversion = speculationAnalyzer.SpeculativeSemanticModel.GetConversion(speculationAnalyzer.ReplacedExpression, cancellationToken);
+        if (conversion.IsIdentity)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs:97]: @cston  this appears to be an error.  Once this merged, i'll open a bug on this.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs [hunk 3]
```diff
@@ -465,6 +449,38 @@ void AddCommaIfMissing()
...
+                {
+                    for (var current = node; current != null; current = current.Parent)
+                    {
+                        if (current is StatementSyntax or MemberDeclarationSyntax)
+                            return current;
+                    }
+                }
+
+                return node;
+            }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs:483]: this improved indentation tracking in a few more situations.

File: src/Features/RulesMissingDocumentation.md [hunk 4]
```diff
@@ -17,6 +17,7 @@ IDE0301 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-r
 IDE0302 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0302> | Simplify collection initialization |
 IDE0303 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0303> | Simplify collection initialization |
 IDE0304 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0304> | Simplify collection initialization |
+IDE0305 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0304> | Simplify collection initialization |
```
Comment (CyrusNajmabadi) [inline src/Features/RulesMissingDocumentation.md:20]:
```suggestion
IDE0305 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0305> | Simplify collection initialization |
```

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForCreateTests.cs [hunk 5]
```diff
@@ -927,4 +926,87 @@ void M()
             ReferenceAssemblies = ReferenceAssemblies.Net.Net80,
         }.RunAsync();
     }
+
+    [Fact]
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForCreateTests.cs:930]: beefed up testing in a few existing features.

File: src/Analyzers/Core/Analyzers/UseCollectionInitializer/UpdateExpressionState.cs [hunk 6]
```diff
@@ -266,6 +284,16 @@ public bool ValuePatternMatches(TExpressionSyntax expression)
...
 
+        var memberAccess = this.SyntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+        if (!this.SyntaxFacts.IsSimpleMemberAccessExpression(memberAccess))
+            return false;
+
+        this.SyntaxFacts.GetPartsOfMemberAccessExpression(memberAccess, out var localInstance, out var memberName);
+        this.SyntaxFacts.GetNameAndArityOfSimpleName(memberName, out var name, out var arity);
+
+        if (arity != 0 || !Equals(name, methodName))
+            return false;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseCollectionInitializer/UpdateExpressionState.cs:295]: moved the fast checks up.

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal.

**Review by akhera99 — APPROVED** [review]

**Pull request merged**